### PR TITLE
all: Allow destructuring into existing variables/ from nested Tuples

### PR DESCRIPTION
**Pull request by robryanx**
If destructure is not preceded by `let`/`mut` existing variables will be
destructured into rather than new variables be declared.

Allow destructuring recursively for nested Tuples.

Introduced `ParsedExpression::Destructure` to be used for the 
destructure pattern to allow easier pattern matching destructures on 
enums in a future PR. Currently it takes the same form as 
`ParsedExpression::JaktTuple`.

Updated naming scheme for the Tuple being destructured to work like
anonymous lambdas to avoid naming ambiguity when the same variables are 
destructured into multiple times.

### CODE REVIEW

File: samples/tuples/destructuring_assignment.jakt [hunk 1]
```diff
@@ -1,11 +1,42 @@
...
+    b++
+    c++
+
+    println("{}, {}, {}", a, b, c)
+
+    let nested_tuple = (1, (2, 3))
+
+    mut test = Test(a: 5, b: 6, c: 7)
+
+    (test.a, (test.b, test.c)) = nested_tuple
```
Comment (alimpfard) [inline samples/tuples/destructuring_assignment.jakt:39]: :thinking: strong python-style "WAT" vibes here, not sure if I'm totally in love with this tbh

File: selfhost/typechecker.jakt [hunk 2]
```diff
@@ -4560,7 +4690,7 @@ struct Typechecker {
         Function(captures, params, can_throw, return_type, block, span) => .typecheck_lambda(captures, params, can_throw, return_type, block, span, scope_id, safety_mode)
         Try(expr, catch_block, catch_name, span) => .typecheck_try(expr, catch_block, catch_name, scope_id, safety_mode, span, type_hint)
         TryBlock(stmt, catch_block, error_name, error_span, span) => .typecheck_try_block(stmt, error_name, error_span, catch_block, scope_id, safety_mode, span)
-        Operator => {
+        else => {
```
Comment (alimpfard) [inline selfhost/typechecker.jakt:4702]: Please don't add an else here, let's just have the compiler yell at us when we add new ones

File: selfhost/typechecker.jakt [hunk 3]
```diff
@@ -3494,19 +3494,28 @@ struct Typechecker {
...
-
-        checked_var_decl = .typecheck_var_decl(var: var_decl, init: expr, scope_id, safety_mode, span)
-        if checked_var_decl! is VarDecl(var_id) {
-            tuple_variable = .program.get_variable(var_id)
+        if checked_expr is Var(var) {
+            let maybe_var = .find_var_in_scope(scope_id, var: var.name)
+            if maybe_var.has_value() {
+                tuple_variable = maybe_var!
+            } else {
+                .error(format("Tuple `{}` not found", var.name), span)
```
Comment (alimpfard) [inline selfhost/typechecker.jakt:3502]: err...`variable {} not found`?

File: selfhost/typechecker.jakt [hunk 4]
```diff
@@ -3494,19 +3494,28 @@ struct Typechecker {
...
+                is_mutable: false,
+                inlay_span: None,
+                span
+            )
+
+            checked_var_decl = .typecheck_var_decl(var: var_decl, init: expr, scope_id, safety_mode, span)
+            if checked_var_decl! is VarDecl(var_id) {
+                tuple_variable = .program.get_variable(var_id)
+            } else {
+                .error(format("Internal destructure varible `{}` not found", var_decl.name), span)
```
Comment (alimpfard) [inline selfhost/typechecker.jakt:3517]: This seems like a strange detail to expose as an error, given that it's not the user's fault, maybe panic instead?

**Review by alimpfard — COMMENTED** [review]
TLDR: Not sure if I love the destructure-into-random-expressions bit, but here's some reviews regardless.

File: samples/tuples/destructuring_assignment.jakt (hunk 1 above)
Comment (robryanx) [reply samples/tuples/destructuring_assignment.jakt:39]: The extension on what we have currently here is specifically for `IndexedStruct`.
For example in the compiler it would allow:
```
namespace_.name = name!.0
namespace_.name_span = name!.1
```
To be rewritten as
```
(namespace_.name, namespace_.name_span) = name!
```
Happy to remove it though if you think it is too much.

Comment (alimpfard) [reply samples/tuples/destructuring_assignment.jakt:39]: We can keep it and see if it does anything I guess, I get the feeling that it's going to be mildly surprising behaviour.

File: samples/tuples/destructuring_assignment.jakt [hunk 5]
```diff
@@ -1,11 +1,48 @@
...
+
+    print("{}, ", a)
+    print("{}, ", b)
+    println("{}", c)
+
+    let nested_tuple = (1, (2, 3))
+
+    mut test = Test(a: 5, b: 6, c: 7)
+
+    (test.a, (test.b, test.c)) = nested_tuple
```
Comment (jntrnr) [inline samples/tuples/destructuring_assignment.jakt:43]: IMHO destructuring should go into bindings not expressions. In the above you're overloading the concept of assignment and binding, but I think you want to keep binding a bit more pure. For example, this would be fine:

```
let (e, (f, g)) = nested_tuple
```

File: tests/typechecker/destructure_missing_variable.jakt [hunk 6]
```diff
@@ -0,0 +1,8 @@
+/// Expect:
+/// - error: "Variable 'bar' not found"
+
+function main() {
+    mut foo = 0
+
+    (foo, bar) = (1, 2)
```
Comment (jntrnr) [inline tests/typechecker/destructure_missing_variable.jakt:7]: Same here, the destructuring should go into bindings not assignments. `foo` already exists so unless you shadow it, you can't rebind it.

File: tests/typechecker/destructure_private.jakt [hunk 7]
```diff
@@ -0,0 +1,19 @@
...
+        return TestClass(x, y)
+    }
+}
+
+function main() {
+    let tuple = (8, 9)
+
+    let test_class = TestClass::make(x: 6, y: 7)
+
+    (test_class.x, test_class.y) = tuple
```
Comment (jntrnr) [inline tests/typechecker/destructure_private.jakt:18]: Same here.

I'll stop repeating it, you get the idea.

**Comment by jntrnr** [comment]
Left some comments. Just to echo some of @alimpfard's thoughts above, I think it's important to keep bindings separate from other types of things (eg expressions or assignments). This makes it easier to read and a bit less surprising.

**Comment by robryanx** [comment]
> Left some comments. Just to echo some of @alimpfard's thoughts above, I think it's important to keep bindings separate from other types of things (eg expressions or assignments). This makes it easier to read and a bit less surprising.

I think Ali was specifically questioning the case on destructing into struct properties (IndexedStruct), which mainly seemed nice to me for completeness and could certainly be removed.

The idea of the PR was to allow for both assignments and bindings so:
`let (a, b) = (3, 4)`
Would we what we have currently.

But we could then have:
```
mut a = 1
mut b = 2

(a, b) = (3, 4)
```
Without a `mut`/`let` keyword for assigning new values to existing variables

The primary use case would be conditional destructing
```
mut a = 1
mut b = 2

if some_condition {
    (a, b) = (3, 4)
}
```

Whereas currently we might do
```
mut a = 1
mut b = 2

if some_condition {
    let (temp_a, temp_b) = (3, 4)

    a = temp_a
    b = temp_b
}
```
or just
```
mut a = 1
mut b = 2

if some_condition {
    let tuple = (3, 4)

    a = tuple.0
    b = tuple.1
}
```

So I guess the question is do we want destructing into existing variables. If so with this syntax or something else?
